crates/biome_js_analyze/src/lint/nursery/use_tan_stack_query_exhaustive_deps.rs

```diff
@@ -0,0 +1,524 @@
+use biome_analyze::{Rule, RuleDiagnostic, RuleSource, context::RuleContext, declare_lint_rule};
+use biome_console::markup;
+use biome_js_semantic::{Binding, SemanticModel};
+use biome_js_syntax::{
+    AnyJsExpression, AnyJsMemberExpression, AnyJsNamedImportSpecifier, AnyJsObjectMember,
+    JsArrayExpression, JsCallExpression, JsIdentifierBinding, JsImport, JsSyntaxKind, JsSyntaxNode,
+};
+use biome_rowan::{AstNode, AstSeparatedList};
+use biome_rule_options::use_tan_stack_query_exhaustive_deps::UseTanStackQueryExhaustiveDepsOptions;
+use rustc_hash::FxHashSet;
+
+use crate::services::semantic::Semantic;
+
+declare_lint_rule! {
+    /// Enforce that all dependencies used in TanStack Query's queryFn are included in the queryKey.
+    ///
+    /// This rule ensures that all variables and dependencies used inside the `queryFn` of TanStack Query
+    /// hooks (like `useQuery` and `useInfiniteQuery`) are properly included in the `queryKey` array.
+    /// This prevents stale closures and ensures proper cache invalidation.
+    ///
+    /// ## Examples
+    ///
+    /// ### Invalid
+    ///
+    /// ```js,expect_diagnostic
+    /// import { useQuery } from '@tanstack/react-query';
+    ///
+    /// function MyComponent({ userId }) {
+    ///   const query = useQuery({
+    ///     queryKey: ['users'],
+    ///     queryFn: () => fetchUser(userId)
+    ///   });
+    /// }
+    /// ```
+    ///
+    /// ### Valid
+    ///
+    /// ```js
+    /// import { useQuery } from '@tanstack/react-query';
+    ///
+    /// function MyComponent({ userId }) {
+    ///   const query = useQuery({
+    ///     queryKey: ['users', userId],
+    ///     queryFn: () => fetchUser(userId)
+    ///   });
+    /// }
+    /// ```
+    ///
+    pub UseTanStackQueryExhaustiveDeps {
+        version: "next",
+        name: "useTanStackQueryExhaustiveDeps",
+        language: "js",
+        sources: &[RuleSource::EslintReactHooks("exhaustive-deps").same()],
```

Comment:
This is not true. Please add `@tanstack/eslint-plugin-query` to the `RuleSource` enum and use it here instead.

Reply:
Best to first pull in main (due to some order changes) before adding this RuleSource to `crates/biome_analyze/src/rule.rs`